src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+import { biginterify, ZERO, ONE, MathUtil } from './utils';
+import { Logger } from '../../types';
+import { PoolState, ReadonlyOrWritable } from './types';
+
+export class NervePoolMath {
+  readonly FEE_DENOMINATOR = biginterify(10 ** 10);
+  readonly A_PRECISION = biginterify(100);
+  readonly MAX_LOOP_LIMIT = biginterify(256);
+  readonly POOL_PRECISION_DECIMALS = biginterify(18);
+
+  constructor(protected name: string, protected logger: Logger) {}
+
+  calculateSwap(
+    state: PoolState,
+    tokenIndexFrom: number,
+    tokenIndexTo: number,
+    dx: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const xp = this._xp(state);
+
+    // uint256 x = dx.mul(self.tokenPrecisionMultipliers[tokenIndexFrom])
+    //    .add(xp[tokenIndexFrom]);
+    const x =
+      dx * state.tokenPrecisionMultipliers[tokenIndexFrom] + xp[tokenIndexFrom];
+
+    const y = this._getY(
+      state,
+      tokenIndexFrom,
+      tokenIndexTo,
+      x,
+      xp,
+      blockTimestamp,
+    );
+
+    if (y === null) {
+      state.isValid = false;
+      return null;
+    }
```

Comment:
When do you get an invalid state?

Reply:
Check the comments below on better handling of this

Reply:
We still have invalid state for unsupported states and when error occured while calculating new state, but it become more readable